data_structures/raster_source.cpp

```diff
@@ -0,0 +1,183 @@
+    {
+        return {};
+    }
+
+    unsigned xthP = (_lon - xmin) / xstep;
+    unsigned ythP = (ymax - _lat) / ystep;
+    int top = floor(ythP);
+    int bottom = ceil(ythP);
+    int left = floor(xthP);
+    int right = ceil(xthP);
+
+    float x = (_lon - left * xstep + xmin) / xstep;
+    float y = (ymax - top * ystep - _lat) / ystep;
+    float x1 = 1.0 - x;
+    float y1 = 1.0 - y;
+
+    return {static_cast<std::int32_t>(
+        raster_data(left, top) * (x1 * y1) + raster_data(right, top) * (x * y1) +
+        raster_data(left, bottom) * (x1 * y) + raster_data(right, bottom) * (x * y))};
+}
+
+// Load raster source into memory
+int SourceContainer::loadRasterSource(const std::string &source_path,
+                                      double xmin,
+                                      double xmax,
+                                      double ymin,
+                                      double ymax,
+                                      unsigned nrows,
+                                      unsigned ncols)
+{
+    int _xmin = static_cast<int>(xmin * COORDINATE_PRECISION);
+    int _xmax = static_cast<int>(xmax * COORDINATE_PRECISION);
+    int _ymin = static_cast<int>(ymin * COORDINATE_PRECISION);
+    int _ymax = static_cast<int>(ymax * COORDINATE_PRECISION);
+
+    auto itr = LoadedSourcePaths.find(source_path);
+    if (itr != LoadedSourcePaths.end())
+    {
+        std::cout << "[source loader] Already loaded source '" << source_path << "' at source_id "
+                  << itr->second << std::endl;
+        return itr->second;
+    }
+
+    int source_id = LoadedSources.size();
+
+    std::cout << "[source loader] Loading from " << source_path << "  ... " << std::flush;
+    TIMER_START(loading_source);
+
+    if (!boost::filesystem::exists(source_path.c_str()))
+    {
+        throw osrm::exception("error reading: no such path");
+    }
+
+    RasterGrid rasterData{source_path, ncols, nrows};
+
+    RasterSource source{rasterData,
+                        static_cast<std::size_t>(ncols),
+                        static_cast<std::size_t>(nrows),
+                        _xmin,
+                        _xmax,
+                        _ymin,
+                        _ymax};
+    TIMER_STOP(loading_source);
+    LoadedSourcePaths.emplace(source_path, source_id);
+    LoadedSources.push_back(std::move(source));
+
+    std::cout << "ok, after " << TIMER_SEC(loading_source) << "s" << std::endl;
+
+    return source_id;
+}
+
+// External function for looking up nearest data point from a specified source
+RasterDatum SourceContainer::getRasterDataFromSource(unsigned int source_id, int lon, int lat)
+{
+    if (LoadedSources.size() < source_id + 1)
+    {
+        throw osrm::exception("error reading: no such loaded source");
+    }
+
+    const auto &found = LoadedSources[source_id];
+    return found.getRasterData(float(lon) / COORDINATE_PRECISION,
+                               float(lat) / COORDINATE_PRECISION);
```

Comment:
`static_cast<T>(x)` instead of old-style `T(x)` C-casts. The reason for this is that C-style casts are implicitely trying "safe casts" (what `static_cast` is doing) and if there is no way to do this, they silently fall back to casts such as `reinterpret_cast` (interpreting bit patterns) and `dynamic_cast` (up/down casting in a type hierarchy), that you really do not want normally.

In addition, C-casts are hard to grep for.


Note: I know OSRM uses C-style casts in a lot of places --- does not make it better though :)